### PR TITLE
Explicitely use python2

### PR DESCRIPTION
**Pull request by lhuard1A**
Some distributions are using python3 as the default python.
On those ones, we need to explicitely use python2.

### CODE REVIEW

**Comment by lhuard1A** [comment]
@twiest : We already discussed that change in PR #46 but I’ll try to split #46 in smaller changes.
I copy my explanation here for reference:

We need it on bleeding edge distributions on which `python` is a symlink to `python3`.
On my Arch Linux, without that patch and with `#!/usr/bin/env python`, I get the following error:
```
    print data_to_print
                      ^
SyntaxError: Missing parentheses in call to 'print'
```

The full story has been discussed on Ansible’s github there:
https://github.com/ansible/ansible/issues/2974

In short, for the distributions where `python` is `python3`,
* the install script will take care of patching the shebang,
* the version of the interpreter used on minions can be overridden thanks to `ansible_python_interpreter`
* But as this file is shipped directly in this repo, it gets no chance to be "fixed".

**Comment by twiest** [comment]
For posterity, it is known that we are changing upstream scripts (ec2.py and gce.py), and as such each time we update those scripts we will need to make these same modifications.

It is also known that the person who's updating the scripts will likely forget to make these changes, and that there will therefore need to be additional PRs filed to fix these.

With those above caveats (and complaints as I hate modifying upstream code like this),

:+1:

**Comment by lhuard1A** [comment]
The point is that today, the script fails on Arch with a rather unhelpful message:

```
$ ./cluster.sh create test

Creating 1 openshift-master instance(s) in AWS...
Clearing inventory cache... Done.
Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 324, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/bin/ansible-playbook", line 160, in main
    inventory = ansible.inventory.Inventory(options.inventory, vault_password=vault_pass)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 118, in __init__
    self.parser = InventoryScript(filename=host_list)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/script.py", line 49, in __init__
    self.groups = self._parse(stderr)
  File "/usr/lib/python2.7/site-packages/ansible/inventory/script.py", line 57, in _parse
    self.raw  = utils.parse_json(self.data)
  File "/usr/lib/python2.7/site-packages/ansible/utils/__init__.py", line 551, in parse_json
    results = json.loads(data)
  File "/usr/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```

The Arch behavior is described in [PEP 394](https://www.python.org/dev/peps/pep-0394/)
Here is an extract of the PEP:
* for the time being, all distributions _should_ ensure that `python` refers to the same target as `python2` .
* however, end users should be aware that `python` refers to `python3` on at least Arch Linux (that change is what prompted the creation of this PEP), so `python` should be used in the shebang line only for scripts that are source compatible with both Python 2 and 3.
* in preparation for an eventual change in the default version of Python, Python 2 only scripts should either be updated to be source compatible with Python 3 or else to use `python2` in the shebang line. 

From Ansible point of view ([issue #2974](https://github.com/ansible/ansible/issues/2974)),
* If Arch is the host from where Ansible is run, `distutils` takes care of patching the shebang at install time.
* If Arch is the minions being configured, `ansible_python_interpreter` can be use to force use of `python2`.


The problem I have here is with `ec2.py` and `gce.py` scripts which are copied from Ansible and checked in here. So, it doesn’t fall in the two above cases.

The problem you describe is coming from the fact that those files are manually copied from outside and checked in.

They should always be automatically fetched from upstream by a piece of script, let’s say at the beginning of `cluster.sh`, instead of being manually re-synchronized from time to time.
That piece of script could take care of making those scripts more portable just after having fetched them.

**Comment by twiest** [comment]
We can't rely on cluster.sh because we do not always run these playbooks using cluster.sh.

Here are the use cases that we use _all the time_.
- Run from cluster.sh
- Run from cloud.rb directly
- Run using ansible-playbook (both OPS and OSE to do it this way)
- Run from tower (near future for OPS)

Therefore we can't have a script download them.

And if cluser.sh did download these, I assume you'd also want it to patch the shebang. Otherwise it'd have the exact same problem these versions have.

Also, if we're automatically fetching them from upstream, we can't guarantee they will be compatible with our scripts. We just updated the gce.py script manually and it broke this command './cloud.rb gce list'.

I wish these were shipped with the ansible RPM, but they simply aren't, so this is the best other way to get everything we want.

Now, all that being said, I did :+1: it above, I just wanted to document for posterity that we are aware of the issues of doing it this way.

**Comment by detiber** [comment]
@twiest eventually, it would be nice to see the dynamic inventory scripts being shipped with the ansible package itself, then we could just ensure that a symlink exists to the latest packaged version without having to manually maintain a copy ourselves.

**Comment by detiber** [comment]
@lhuard1A What about a step to have cluster.sh (or even cloud.rb) detect that it is being run on a host that doesn't have python2 as the default python, then it could verify that python2 is installed and either manipulate the path variable to set the python2 path before the system python path, or it could create and instantiate a virtualenv for python2?

**Comment by lhuard1A** [comment]
@detiber : Thanks for the suggestion about virtualenv!
I implemented it in 361d5ea. Could you check if it is well what you had in mind?
It has the advantage of making `cluster.sh` work for me without requiring to patch a lot of files.
It has the disadvantage of fixing only `cluster.sh`.
@twiest pointed out the fact that people may also call `cloud.rb` or `ansible-playbook` directly. These use cases are not fixed.

**Comment by twiest** [comment]
@lhuard1A does that mean you no longer need this PR?

**Comment by lhuard1A** [comment]
@twiest : I need this PR.
If don’t like the idea of patching the shebang of the `gce.py` and `aws.py` files and reject this PR, I can open another PR (or amend this one) with the virtualenv stuff proposed by @detiber in order to have 361d5ea merged upstream.
I would prefer to have this PR merged because it fixes all the use cases (use of `cluster.sh`, `cloud.rb` and `ansible-playbook`) whereas the virtualenv stuff fixes only the use of `cluster.sh`.

**Comment by detiber** [comment]
@lhuard1A @twiest I think I found a better solution...  Ansible has an environment variable to specify the python path for hosts that do not have /usr/bin/python default to python2. Setting ansible_python_interpreter to the path of the python2 binary should be sufficient for the ansible-playbook method. http://docs.ansible.com/intro_inventory.html#list-of-behavioral-inventory-parameters

Also, we should probably honor the same variable within the ansible-helper script to invoke any python scripts using the specified interpreter.

**Comment by lhuard1A** [comment]
@detiber My understanding was that `ansible_python_interpreter` was only for the python interpreter to use on remote hosts.
Let’s say that I’m on `my_workstation` where I run `ansible` to provision the machine `my_remote_server`.

    Machine
    my_workstation     1.  ansible-playbook                 # python interpreter chosen at ansible installation
    my_workstation     2.  +-> lib/aws.py                   # problematic
    my_workstation     3.  +-> ssh my_remote_server python ~/.ansible/tmp/…
    my_remote_server   4.      +-> python ~/.ansible/tmp/…  # python interpreter chosen by ansible_python_interpreter

1. `ansible-playbook` is run from `my_workstation`. I can rely on the packager of my distribution to have installed ansible so that it works seamlessly on my machine. Concretely, distutils will take care of putting an appropriate shebang on what it installs.
3. `ansible` works by pushing scripts to the provisioned machine and executing them via ssh.
4. The remote script runs on `my_remote_server` over ssh. The python interpreter to use on `my_remote_server` can be customized thanks to the `ansible_python_interpreter` variable attached to that host in the inventory.

If `my_remote_server` was a machine with the default python pointing to python3, we could have set `ansible_python_interpreter` on `my_remote_server` to point to a python2.
In our case, our roles/playbooks are very “RedHat based distribution oriented” (Use of `yum`, `etc/sysconfig` layout, etc.). And on those “RedHat based distributions” we don’t have that issue.

The problem is mostly for 2. `lib/aws.py`. It is run of `my_workstation`. So, it is not affected by `ansible_python_interpreter`. It is not either installed with the `ansible` package.

**Comment by detiber** [comment]
@lhuard1A ah, you are indeed correct. I'm good with changing the #! line for now, I think we'll need to tackle it more directly in the future though. Especially since we'll be replacing cloud.rb with a python script.

**Comment by detiber** [comment]
@twiest +1 to merging this

**Comment by twiest** [comment]
:+1:  with the caveats I listed previously.
